scripts/pylib/pytest-twister-harness/src/twister_harness/device/hardware_adapter.py

```diff
@@ -0,0 +1,202 @@
+        """
+        Open serial connection.
+
+        :param timeout: Read timeout value in seconds
+        """
+        if self.connection:
+            # already opened
+            return
+
+        serial_name = self._open_serial_pty() or self.device_config.serial
+        logger.info('Opening serial connection for %s', serial_name)
+        try:
+            self.connection = serial.Serial(
+                serial_name,
+                baudrate=self.device_config.baud,
+                parity=serial.PARITY_NONE,
+                stopbits=serial.STOPBITS_ONE,
+                bytesize=serial.EIGHTBITS,
+                timeout=timeout
+            )
+        except serial.SerialException as e:
+            logger.exception('Cannot open connection: %s', e)
+            self._close_serial_pty()
+            raise
+
+        self.connection.flush()
+
+    def disconnect(self) -> None:
+        """Close serial connection."""
+        if self.connection:
+            serial_name = self.connection.port
+            self.connection.close()
+            self.connection = None
+            logger.info('Closed serial connection for %s', serial_name)
+        self._close_serial_pty()
+
+    def _open_serial_pty(self) -> str | None:
+        """Open a pty pair, run process and return tty name"""
+        if not self.device_config.serial_pty:
+            return None
+        master, slave = pty.openpty()
+        try:
+            self.serial_pty_proc = subprocess.Popen(
+                re.split(',| ', self.device_config.serial_pty),
+                stdout=master,
+                stdin=master,
+                stderr=master
+            )
+        except subprocess.CalledProcessError as e:
+            logger.exception('Failed to run subprocess %s, error %s',
+                             self.device_config.serial_pty, str(e))
+            raise
+        return os.ttyname(slave)
+
+    def _close_serial_pty(self) -> None:
+        """Terminate the process opened for serial pty script"""
+        if self.serial_pty_proc:
+            self.serial_pty_proc.terminate()
+            self.serial_pty_proc.communicate()
+            logger.info('Process %s terminated', self.device_config.serial_pty)
+            self.serial_pty_proc = None
+
+    def generate_command(self) -> None:
+        """Return command to flash."""
+        west = shutil.which('west')
+        if west is None:
+            raise TwisterHarnessException('west not found')
+
+        command = [
+            west,
+            'flash',
+            '--skip-rebuild',
+            '--build-dir', str(self.device_config.build_dir),
+        ]
+
+        command_extra_args = []
+        if self.device_config.west_flash_extra_args:
+            command_extra_args.extend(self.device_config.west_flash_extra_args)
+
+        if runner := self.device_config.runner:
+            command.extend(['--runner', runner])
+
+            if board_id := self.device_config.id:
```

Comment:
as discussed already, this is a major flaw if we have to maintain such code twice, we need to find a better and only 1 place to implement this logic so there will not be any need to change this code here whenever we add a new flasher or platform specific tooling.

So please add this as something to be resolved in the tracking issue.